src/ripple/core/SQLInterface.h

```diff
@@ -0,0 +1,247 @@
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_SQLINTERFACE_H_INCLUDED
+#define RIPPLE_CORE_SQLINTERFACE_H_INCLUDED
+
+#include <ripple/app/ledger/Ledger.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/app/misc/NetworkOPs.h>
+#include <ripple/app/misc/Transaction.h>
+#include <ripple/core/Config.h>
+#include <ripple/core/DatabaseCon.h>
+#include <ripple/peerfinder/impl/Store.h>
+#include <boost/filesystem.hpp>
+#include <boost/optional.hpp>
+#include <boost/variant.hpp>
+
+namespace ripple {
+
+struct SQLLedgerInfo
+{
+    boost::optional<std::string> sLedgerHash, sPrevHash, sAccountHash,
+        sTransHash;
+    boost::optional<std::uint64_t> totDrops, closingTime, prevClosingTime,
+        closeResolution, closeFlags, ledgerSeq64;
+};
+
+class SQLInterface
+{
+public:
+    enum DatabaseType : uint8_t { ledgerDb, transactionDb, lastDb };
+
+    enum TableType { ledgers, transactions, accountTransactions, lastTable };
+
+    typedef std::pair<LedgerIndex, uint8_t> DatabaseIndex;
+
+    static std::string
+    tableName(TableType type)
+    {
+        switch (type)
+        {
+            case ledgers:
+                return "Ledgers";
+            case transactions:
+                return "Transactions";
+            case accountTransactions:
+                return "AccountTransactions";
+            case lastTable:
+                static_assert(lastTable == 3);
+        }
+        return "Unknown";
+    }
+
+    struct RowsMinMax {
+        int rows;
+        int min;
+        int max;
+    };
+
+    struct Hashes {
+        uint256 ledgerHash;
+        uint256 parentHash;
+    };
+
+    static std::unique_ptr<SQLInterface>
+    init(Application& app, Config const& config);
+
+    virtual bool
+    makeLedgerDBs(
+        Config const& config,
+        DatabaseCon::Setup const& setup,
+        DatabaseCon::CheckpointerSetup const& checkpointerSetup) = 0;
+
+    virtual void
+    addDBs(LedgerIndex shardIndex, DatabaseCon* tx, DatabaseCon* lgr) = 0;
+
+    virtual void
+    removeDBs(LedgerIndex shardIndex) = 0;
```

Comment:
I think the shard specific functions should be moved to a derived class.

Reply:
Done.